ui/components/app/simulation-details/useSimulationMetrics.ts

```diff
@@ -0,0 +1,265 @@
+import {
+  UseDisplayNameRequest,
+  UseDisplayNameResponse,
+  useDisplayNames,
+} from '../../../hooks/useDisplayName';
+import { NameType } from '@metamask/name-controller';
+import { MetaMetricsContext } from '../../../contexts/metametrics';
+import {
+  MetaMetricsEventCategory,
+  MetaMetricsEventName,
+} from '../../../../shared/constants/metametrics';
+
+export type UseSimulationMetricsProps = {
+  balanceChanges: BalanceChange[];
+  loadingTime?: number;
+  simulationData?: SimulationData;
+  transactionId: string;
+};
+
+export enum SimulationResponseType {
+  Failed = 'failed',
+  Reverted = 'transaction_revert',
+  NoChanges = 'no_balance_change',
+  Changes = 'balance_change',
+  InProgress = 'simulation_in_progress',
+}
+
+export enum AssetType {
+  Native = 'native',
+  ERC20 = 'erc20',
+  ERC721 = 'erc721',
+  ERC1155 = 'erc1155',
+}
+
+export enum FiatType {
+  Available = 'available',
+  NotAvailable = 'not_available',
+}
+
+export enum PetnameType {
+  Saved = 'saved',
+  Default = 'default',
+  Unknown = 'unknown',
+}
+
+export function useSimulationMetrics({
+  balanceChanges,
+  loadingTime,
+  simulationData,
+  transactionId,
+}: UseSimulationMetricsProps) {
+  const dispatch = useDispatch();
+
+  const displayNameRequests: UseDisplayNameRequest[] = balanceChanges.map(
+    ({ asset }) => ({
+      value: asset.address ?? '',
+      type: NameType.ETHEREUM_ADDRESS,
+      preferContractSymbol: true,
+    }),
+  );
+
+  const displayNames = useDisplayNames(displayNameRequests);
+
+  const displayNamesByAddress = displayNames.reduce(
+    (acc, displayNameResponse, index) => ({
+      ...acc,
+      [balanceChanges[index].asset.address ?? '']: displayNameResponse,
+    }),
+    {} as { [address: string]: UseDisplayNameResponse },
+  );
+
+  const { updateTransactionEventFragment } = useTransactionEventFragment();
+
+  useIncompleteAssetEvent(balanceChanges, displayNamesByAddress);
+
+  const receivingAssets = balanceChanges.filter(
+    (change) => !change.amount.isNegative,
+  );
+
+  const sendingAssets = balanceChanges.filter(
+    (change) => change.amount.isNegative,
+  );
+
+  const simulation_response = getSimulationResponseType(simulationData);
+  const simulation_latency = loadingTime;
+
+  const properties = {
+    simulation_response,
+    simulation_latency,
+    ...getProperties(
+      receivingAssets,
+      'simulation_receiving_assets_',
+      displayNamesByAddress,
+    ),
+    ...getProperties(
+      sendingAssets,
+      'simulation_sending_assets_',
+      displayNamesByAddress,
+    ),
+  };
+
+  const sensitiveProperties = {
+    ...getSensitiveProperties(receivingAssets, 'simulation_receiving_assets_'),
+    ...getSensitiveProperties(sendingAssets, 'simulation_sending_assets_'),
+  };
+
+  useEffect(() => {
+    updateTransactionEventFragment(
+      { properties, sensitiveProperties },
+      transactionId,
+    );
+  }, [dispatch, transactionId, JSON.stringify(properties)]);
```

Comment:
Should we just remove `dispatch` here?
Also what's the reason of using `JSON.stringify` here?

Reply:
Good spot, `dispatch` was used in an older implementation, removed.

The `JSON.stringify` is a means to use a deep equal comparison on the parameters as a dependency to the effect, so we don't update unless the params themselves have changed. It requires primitive values hence this provides it as a string.

I could follow the entire rendering stack backwards and use `useMemo` etc but ultimately none of it is performance heavy so this is just a final guard to limit how often we update via the metrics controller.

Reply:
One more small concern, `JSON.stringify` does not guarantee the order of properties in the serialized string. If the properties of the object are the same but their order changes, may produce a different string, causing the effect to re-run unnecessarily.

Reply:
I don't believe it guarantees any specific order, but it does provide a consistent / stable order. Meaning the same object should serialize the same way consistently which is what we need.